text/0023-starlark_load_dataset.md

```diff
@@ -0,0 +1,249 @@
+
+For these reasons, we should make datasets a first class citizen in starlark scripts by creating a global `load_dataset` statement that can be expanded with permissions scopes in the future. This statement should return a _dataset document_ that we can expand the functionality of over time. 
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+### `load_dataset`
+This RFC proposes switching to a single method of loading a dataset that is analogous to loading a module. Dataset dependencies will be declared (by convention) grouped together at the top of a script, forming a clear delaration of the datasets this script depends on: 
+
+```python
+load("http.star", "http")
+# load a dataset into a variable named "fhv"
+fhv = load_dataset("b5/nyc_for_hire_vehicles")
+
+def download(ctx):
+  # use the fhv dataset to inform an http request
+  vins = ["%s,%s" % (entry['vin'], entry['model_yearl']), for entry in fhv.body()]
+
+  res = http.post("https://vpic.nhtsa.dot.gov/api/vehicles/DecodeVINValuesBatch/", form_body={
+    'format': 'JSON', 
+    'DATA': vins.join(";")
+  })
+
+  return res.json()
+
+def transform(ds, ctx):
+  ds.set_body(ctx.download)
+```
+
+In this example, the `load_dataset` statement supplies no optional arguments, it's assumed this dataset is loaded and available in all parts of the script. This RFC defines no optional arguments for `load_dataset`, so loading datasets for global use is the only option available. Considering all datasets on Qri are currently public, this is a sensible place to start.
+
+### Access Control
+Knowing that things like "private datasets" are planned for the future, we need an explicit way to load a dataset that we can build upon. While we can't yet define what those nuances will be, we do need to define _how they will be expressed_. Here are examples of possible scoping statements, expressed as optional arguments: 
+
+```python
+# scope a dataset load to only meta & structure components
+residents = load_dataset("b5/city_residents", components=["md", "st"])
+
+# only allow access to "residents" during the "transform" step. "residents" will 
+# be equal to None during all other steps
+residents = load_dataset("b5/city_residents", steps=["transform"])
+
+# apply differential privacy to all statements that read from the body
+residents = load_dataset("b5/city_residents", anonymized=True)
+
+# require that the permissions of the dataset that results from this transform
+# match the permissions of "b5/city_residents"
+residents = load_dataset("b5/city_residents", forward_permissions=True)
+```
+
+These are suggestions only. We need lots of time to think through how these options will work, this rfc only adds the assumption that those statements will be expressed as options on `load_dataset`. We can build on the already-present assumption that loaded datasets are read-only, making all options pertain to read access.
```

Comment:
Important to note: these options are not giving additional permissions to scripts, as that would go against the no-need-to-audit goal, they are instead self-restricting what the loaded dataset may be used for. They exist for the benefit of the transform author, not for the security of the transform consumer.